### DISCUSSION TITLE
Does xarray automatically use Opportunistic Caching from Dask?

### DISCUSSION THREAD

**Discussion by ahuang11** [pydata/xarray#4580] [Q&A] [answered]
When I do something like
```
ds = xr.tutorial.open_dataset('air_temperature').chunk()
ds['max'] = ds['air'].max()
ds['min'] = ds['air'].min()
ds = ds.compute()
```
Does it automatically do something like
`dd.compute(df.amount.max(), df.amount.min())` so that Dask will share intermediates intelligently, reading through the dataset only once?

https://docs.dask.org/en/latest/caching.html

**Comment by dcherian** [comment, marked as answer]
> Does it automatically do something like
dd.compute(df.amount.max(), df.amount.min()) so that Dask will share intermediates intelligently, reading through the dataset only once?

Yes. The graph for that dataset is computed all at once so intermediaries are shared, and data is only read once.

This is not caching though. For that you need to turn on the cache: https://docs.dask.org/en/latest/caching.html#automatic-opportunistic-caching and then you may see speedups when repeatedly accessing the same data.
